src/streamlink_cli/main.py

```diff
@@ -76,19 +76,25 @@ def create_output(plugin):
      - A regular file
 
     """
+    global formatted_filename
 
     if (args.output or args.stdout) and (args.record or args.record_and_pipe):
         console.exit("Cannot use record options with other file output options.")
 
+    for unformatted_filename in args.output, args.record, args.record_and_pipe:
+        if unformatted_filename and isinstance(unformatted_filename, str):
+            formatted_filename = create_filename(plugin, unformatted_filename)
+            break
+
```

Comment:
What's the reason for this? Even if `args.output`, `args.record` and `args.record_and_pipe` are mutually exclusive, taking the name from the first one that exists doesn't make sense. Is it because of the global variable and the other usages of `args.output` and `args.record` down below?

Reply:
My thinking was that as `args.output`, `args.record` and `args.record_and_pipe` are mutually exclusive and also used separately in a way that appeared to be unnecessary where the formatted name is substituted instead, it would be better to create just the one global variable for a formatted name as a solution, rather than one for each and leave the other usage of any of those three attributes in the code alone.

Reply:
What are your thoughts on replacing `args.output`, `args.record` and `args.record_and_pipe` here?  Do you think there should be separate format processed versions for each?  I'm wondering if a better solution might be to write a class that has these as properties that are either formatted or otherwise set to whatever they were originally.  That class could also then contain a property with the formatted title without the filesystem-safe processing for use where `title` is used currently.  That would mean `create_title()` would be removed and replaced with the class instance, although this would have to be done earlier in various places in the code to make the filename properties available when needed.  The class instance would be assigned to a global.  Does that seem like a reasonable way to go?

Reply:
> What are your thoughts on replacing args.output, args.record and args.record_and_pipe here?

The problem with the global var is that you're setting it in a deeply nested function call while reading it afterwards up the call stack. I don't think this is good. Just check the `args` attributes individually and override their values if a value exists. Ideally this should be done as early as possible, eg. in `setup_args`, but I don't think it's possible in this situation.

As I've mentioned multiple times, the `streamlink_cli` module is in desperate need of a complete rewrite that cleans up the usage of global variables. And it's way too bloated for a single module as well. This is why adding stuff like this makes it unnecessarily difficult :(

> Do you think there should be separate format-processed versions for each?

You mean LazyFormatter instances or different sets of available vars?

Reply:
> The problem with the global var is that you're setting it in a deeply nested function call while reading it afterwards up the call stack. I don't think this is good. Just check the args attributes individually and override their values if a value exists. Ideally this should be done as early as possible, eg. in setup_args, but I don't think it's possible in this situation.

I think maybe I misunderstood what you meant in: https://github.com/streamlink/streamlink/pull/3913#discussion_r688263277.  I was under the impression you were objecting to overriding the three global file args at all and meant separate global vars should be used instead?  Is it the case that you meant overriding is OK, but just not at that point in the code?  It would appear that the earliest point it can be done is in `handle_url()`.

Reply:
Hm. After thinking about it again, setting the `args` attributes like you did in #3913 might be the best solution for now. When I wrote the comment I wasn't 100% aware of how the attributes were read in the following methods. Let's not make it unnecessarily difficult here. The problem of relying on side-effects from other methods when reading the `args` attributes remains though, but that shouldn't be too big of a deal for this feature. If I start arguing about that then I need to argue about a lot of other stuff in the CLI module, which I don't want now. Sorry about the confusion.

Reply:
OK.  I did have another idea, otherwise, if setting the args in `handle_url` made sense.  This would have involved adding a `title` keyword arg to the `output_*()` and `create_output()` methods.  The reason for this being that I'd generate the formatted versions of the file args and the title at the same time.  Each of the `plugin.get_*()` methods would only be called once here and only if callable, otherwise taking the default value string.  I don't know if that seems reasonable or you'd rather stick with how it was done in #3913?

If you still prefer the idea of what was done in #3913, I can go back to something like that, although I still like the idea of integrating some of what I've said above, such that the `plugin.get_*()` methods are only called once - if callable - to eventually generate the formatted versions of both the file arg and title.

Either way, I'm guessing that limiting the number of calls to these plugin methods would be considered beneficial to potentially limit any delay in the stream start, as per your comment in this PR (https://github.com/streamlink/streamlink/pull/3933#discussion_r693544988)?

Sorry to throw more confusion into the mix!

Reply:
> `handle_url`

You can't read the stream metadata before the actual stream has been chosen, and this does happen quite late in `output_stream` when `success_open` is `True`, so passing another parameter to the various `output_` methods is not possible.

What looks like a better solution is reading the `filename` attribute of the `FileOutput` instance (if the file descriptor for stdout is not set) instead of overriding the `args` attributes. `args.output` and `args.record` are read in `output_stream` itself and `read_stream`, which is called by `output_stream` and the FileOutput instance is already passed as a parameter.

Reply:
> You can't read the stream metadata before the actual stream has been chosen, and this does happen quite late in output_stream when success_open is True, so passing another parameter to the various output_ methods is not possible.

I'm trying to understand why that's the case.  The `_get_streams` method for the plugin has already been called in `fetch_streams()` via `plugin.streams()`, so it seems like there should be valid available metadata here?  Testing seems to indicate that this is the case?

Reply:
Sorry, I worded this wrong and I should've been more precise. I meant the output type choice.

> The `_get_streams` method for the plugin has already been called

It's up to the plugin's `get_{author,category,title}` methods to return the metadata, which means there could be fetching logic in these methods instead of `_get_streams`, and fetching metadata when it's not necessary should be avoided to prevent unnecessary stream launch delays. Most plugins currently fetch metadata in `_get_streams` and return the cached data in these metadata getter methods, but that's not a guaranteed pattern.

What's important is the output type (passthrough, http, json, stream-url, player/file) that gets decided in `handle_stream`, which means it would only make sense reading metadata in `output_stream`, as this method is responsible for creating the output (player or file) where we need the substituted file name. The `output_stream` call is however part of a for-loop, which depends on its return value, which again depends on its inner `success_open` variable, and this must be `True`. Since you need the substituted file name in the `console.exit` and `progress` calls after the `create_output()` call, it probably makes more sense getting the metadata in `output_stream` rather than `create_output`, or you'd have to return a tuple in `create_output`.

Hope that makes sense and I hope I haven't missed something myself. It's very difficult to keep all potential cases in your head at the same time when the CLI design isn't that great.

Reply:
Many thanks for the clarification.  I think what I have in the third iteration is something quite a bit nicer.  It lazy loads and caches title/author/category/game/url for both title and filename creation, so for example, plugin.get_title() will only be called if there's a title formatting variable in the title and/or filename spec, but it'll only be called once, even if used in both specs.  I know the URL doesn't really need lazy loading/caching, but it's a harmless side effect of the implementation.  Let me put it all together and check that it works as aspected, then I'll push it for you to take a look at it.  I won't bother with the tests until the basic methodology meets with approval.